json/marshal.go

```diff
@@ -0,0 +1,193 @@
+package marshal
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+const FORCE_SEND_FIELD_NAME = "ForceSendFields"
+
+// Marshal returns a JSON encoding of the given object. Included fields:
+// - non-empty value
+// - a basic type whose field's name is present in forceSendFields
+// Our templates always populate either omitempty or the '-' tag.
+// For simplicity, we assume one of those cases.
+func Marshal(object interface{}) ([]byte, error) {
+
+	dataMap, err := structAsMap(object)
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(dataMap)
+}
+
+// Converts the object to a map
+func structAsMap(object interface{}) (map[string]interface{}, error) {
+	result := make(map[string]interface{})
+
+	// If the object is nil or a pointer to nil, don't do anything
+	if object == nil || (reflect.ValueOf(object).Kind() == reflect.Ptr && reflect.ValueOf(object).IsNil()) {
+		return result, nil
+	}
+
+	forceSendFields, err := getForceSendFields(object)
+
+	if err != nil {
+		return nil, err
+	}
+
+	includeFields := make(map[string]bool)
+	for _, field := range forceSendFields {
+		includeFields[field] = true
+	}
+
+	value := reflect.ValueOf(object)
+	value = reflect.Indirect(value)
+	objectType := value.Type()
+
+	for i := 0; i < value.NumField(); i++ {
+		jsonTag := objectType.Field(i).Tag.Get("json")
+		tag, err := parseJSONTag(jsonTag)
+		if err != nil {
+			return nil, err
+		}
+
+		fieldValue := value.Field(i)
+		fieldStruct := objectType.Field(i)
+
+		// Anonymous fields should be marshalled using the same JSON, and then merged into the same map
+		if fieldStruct.Anonymous && fieldValue.IsValid() {
+			dataMap2, err := structAsMap(fieldValue.Interface())
+			if err != nil {
+				return nil, err
+			}
+			result = mergeMaps(result, dataMap2)
+			continue
+		}
+
+		// Skip fields which should not be included
+		if !includeField(tag, fieldValue, fieldStruct, includeFields) {
+			continue
+		}
+
+		if tag.asString {
+			result[tag.name] = formatAsString(fieldValue, fieldStruct.Type.Kind())
+		} else {
+			result[tag.name] = fieldValue.Interface()
+		}
+	}
+	return result, nil
+}
+
+// returns the element as string
+func formatAsString(v reflect.Value, kind reflect.Kind) string {
+	if kind == reflect.Ptr && !v.IsNil() {
+		v = v.Elem()
+	}
+
+	return fmt.Sprintf("%v", v.Interface())
+}
+
+type jsonTag struct {
+	name      string
+	asString  bool
+	ignore    bool
+	omitempty bool
+}
+
+func parseJSONTag(raw string) (jsonTag, error) {
+	if raw == "-" || raw == "" {
+		return jsonTag{ignore: true}, nil
+	}
+
+	parts := strings.Split(raw, ",")
+
+	jsonTag := jsonTag{
+		name: parts[0],
+	}
+
+	for _, v := range parts {
+		switch v {
+		case "omitempty":
+			jsonTag.omitempty = true
+		case "string":
+			jsonTag.asString = true
+		}
+	}
+	return jsonTag, nil
+}
```

Comment:
Does this come from `encoding/json`? If not, is it different?

Reply:
`encoding/json` does the same in a different way:
https://github.com/golang/go/blob/master/src/encoding/json/tags.go
That particular class is not exported, so either we reimplement or copy the file. Since I had already reimplemented it before looking at it, I left my version to avoid licensing discussions.